### PR TITLE
BLOC Management for HomePage 

### PR DESCRIPTION
**Pull request by UrosLITS**
Made changes and added download books and upload books functions as well as mandatory classes for further usage in project

### CODE REVIEW

File: lib/app_routes/app_routes.dart [hunk 1]
```diff
@@ -1,22 +1,26 @@
...
 import 'package:flutter/material.dart';
 import 'package:flutter/widgets.dart';
 
-const String homeRoute = '/';
-const String loginRoute = '/Login';
-const String registerRoute = '/Register';
+const String kHomeRoute = '/';
+const String kLoginRoute = '/Login';
+const String kRegisterRoute = '/Register';
+const String kAddNewBookPage = '/AddNewBookPage';
```
Comment (petarLits) [inline lib/app_routes/app_routes.dart:11]: make better name, it look's like adding book's page, and all other constants are with Route at the end, but this one is with Page

**Review by petarLits — CHANGES_REQUESTED** [review]

File: lib/models/book.dart [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    required this.author,
+    required this.title,
+    required this.url,
+    required this.id,
+  });
+
+  String title;
+  String author;
+  @JsonKey(name: "image")
+  String url;
```
Comment (lecicdjuro) [inline lib/models/book.dart:16]: url can by anything, so please rename it to imageUrl

File: lib/models/book.dart [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  String id;
+
+  //yet to be added
+  //BookData? bookData;
```
Comment (lecicdjuro) [inline lib/models/book.dart:20]: If this class does not exists yet, then remove this commented section

File: lib/presentation/book/add_new_book_page.dart [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+      ),
+      body: Container(
+          padding: EdgeInsets.all(16),
+          child: Form(
+            key: _formKey,
+            child: Column(
+              crossAxisAlignment: CrossAxisAlignment.stretch,
+              children: [
+                SizedBox(height: 30),
+                TextFormField(
```
Comment (lecicdjuro) [inline lib/presentation/book/add_new_book_page.dart:58]: should we use our custom textformfield?

File: lib/presentation/book/add_new_book_page.dart [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+                  ),
+                  validator: (value) {
+                    if (value == null || value.isEmpty) {
+                      return AppLocalizations.of(context)!.empty_book_title;
+                    }
+                    return null;
+                  },
+                ),
+                SizedBox(height: 40),
+                TextFormField(
```
Comment (lecicdjuro) [inline lib/presentation/book/add_new_book_page.dart:90]: ditto

File: lib/presentation/book/add_new_book_page.dart [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+        child: FloatingActionButton(
+          child: Icon(CupertinoIcons.camera),
+          onPressed: () async {
+            final XFile? result =
+                await _imagePicker.pickImage(source: ImageSource.gallery);
+            if (result != null) {
+              imageFile = File(result.path);
+              imageName = imageFile?.path.split("/").last;
+
+              setState(() {});
```
Comment (lecicdjuro) [inline lib/presentation/book/add_new_book_page.dart:189]: we should use bloc to emit state

File: lib/presentation/book/bloc/home_page_bloc.dart [hunk 7]
```diff
@@ -0,0 +1,9 @@
+import 'package:book/presentation/book/bloc/home_page_event.dart';
+import 'package:book/presentation/book/bloc/home_page_state.dart';
+import 'package:flutter_bloc/flutter_bloc.dart';
+
+class HomePageBloc extends Bloc<HomePageEvent, HomePageState> {
+  HomePageBloc() : super((InitialState())) {
+    //on<AddBook>(_onBookAdded);
```
Comment (lecicdjuro) [inline lib/presentation/book/bloc/home_page_bloc.dart:7]: remove dead code

File: lib/presentation/book/bloc/home_page_event.dart [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  final String author;
+  final String title;
+  final String url;
+  final String id;
+
+  AddBook(
+      {required this.title,
+      required this.author,
+      required this.url,
+      required this.id});
```
Comment (lecicdjuro) [inline lib/presentation/book/bloc/home_page_event.dart:17]: format

File: lib/presentation/book/bloc/home_page_event.dart [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+  @override
+  List<Object?> get props => [title, author, url, id, Random().nextInt(10000)];
```
Comment (lecicdjuro) [inline lib/presentation/book/bloc/home_page_event.dart:20]: why Random here?

File: lib/presentation/book/view/home_page_view.dart [hunk 10]
```diff
@@ -9,8 +17,139 @@ class HomePageView extends StatefulWidget {
...
+      ),
+      floatingActionButton: Visibility(
+        visible: appUser.isAdmin == true,
+        child: FloatingActionButton(
+          onPressed: () async {
+            final Book? result =
+                await Navigator.pushNamed<dynamic>(context, kAddNewBookPage);
+
+            if (result != null) {
+              bookList.add(result);
```
Comment (lecicdjuro) [inline lib/presentation/book/view/home_page_view.dart:74]: use bloc to emit new state

File: lib/presentation/book/view/add_new_book_page.dart [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+                          maxLines: 1,
+                        ),
+                      ),
+                      Visibility(
+                        visible: imageFile != null,
+                        child: IconButton(
+                          onPressed: () {
+                            imageFile = null;
+                            imageName = "";
+                            setState(() {});
```
Comment (petarLits) [inline lib/presentation/book/view/add_new_book_page.dart:127]: Use bloc

File: lib/presentation/book/view/add_new_book_page.dart [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+              child: Icon(CupertinoIcons.camera),
+              onPressed: () async {
+                final XFile? result =
+                    await _imagePicker.pickImage(source: ImageSource.gallery);
+                if (result != null) {
+                  imageFile = File(result.path);
+                  imageName = imageFile?.path.split("/").last;
+
+                  setState(() {});
+                }
```
Comment (petarLits) [inline lib/presentation/book/view/add_new_book_page.dart:176]: Use bloc

**Review by petarLits — CHANGES_REQUESTED** [review]

File: lib/data/firebase_db_manager.dart [hunk 13]
```diff
@@ -0,0 +1,26 @@
+import 'dart:io';
+import 'package:book/core/constants.dart';
+import 'package:cloud_firestore/cloud_firestore.dart';
+import 'package:firebase_storage/firebase_storage.dart';
+
+class FirebaseDbManager {
```
Comment (lecicdjuro) [inline lib/data/firebase_db_manager.dart:6]: empty line

File: lib/data/firebase_db_manager.dart [hunk 14]
```diff
@@ -0,0 +1,26 @@
+import 'dart:io';
+import 'package:book/core/constants.dart';
+import 'package:cloud_firestore/cloud_firestore.dart';
+import 'package:firebase_storage/firebase_storage.dart';
+
```
Comment (lecicdjuro) [inline lib/data/firebase_db_manager.dart:5]: make Singleton class

File: lib/presentation/book/view/add_new_book_page.dart [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+  String? imageUrl;
+  late ImagePicker _imagePicker;
+  File? imageFile;
+
+  String? imageName;
+
+  @override
+  void initState() {
+    _imagePicker = ImagePicker();
+    imageName = "";
```
Comment (lecicdjuro) [inline lib/presentation/book/view/add_new_book_page.dart:36]: why?

File: lib/presentation/book/view/add_new_book_page.dart [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+          floatingActionButton: Padding(
+            padding: const EdgeInsets.only(bottom: 80),
+            child: FloatingActionButton(
+              child: Icon(CupertinoIcons.camera),
+              onPressed: () async {
+                final XFile? result =
+                    await _imagePicker.pickImage(source: ImageSource.gallery);
+                if (result != null) {
+                  imageFile = File(result.path);
+                  imageName = imageFile?.path.split("/").last;
```
Comment (lecicdjuro) [inline lib/presentation/book/view/add_new_book_page.dart:173]: make some FileUtils class that will extract name from the file

File: lib/presentation/book/view/home_page_view.dart [hunk 17]
```diff
@@ -9,8 +18,138 @@ class HomePageView extends StatefulWidget {
 }
 
 class _HomeBookPageView extends State<HomePageView> {
+  late final _controller;
+  late List<Book> bookList = [];
+
+  @override
+  void initState() {
+    _controller = PageController(viewportFraction: 0.9, initialPage: 0);
```
Comment (lecicdjuro) [inline lib/presentation/book/view/home_page_view.dart:26]: where we call dispose?

File: lib/presentation/book/view/home_page_view.dart [hunk 18]
```diff
@@ -9,8 +18,138 @@ class HomePageView extends StatefulWidget {
...
+          ),
+          floatingActionButton: Visibility(
+            visible: true,
+            child: FloatingActionButton(
+              onPressed: () async {
+                final Book? result = await Navigator.pushNamed<dynamic>(
+                    context, kAddNewBookRoute);
+
+                if (result != null) {
+                  bookList.add(result);
```
Comment (lecicdjuro) [inline lib/presentation/book/view/home_page_view.dart:70]: It's not a good place to add to the list. The list should be added after the refresh

File: lib/presentation/login/view/login_page_view.dart [hunk 19]
```diff
@@ -38,7 +39,10 @@ class _LoginPageState extends State<LoginPage> {
     return BlocConsumer<LoginBloc, LoginState>(
       listener: (context, state) async {
         if (state is SuccessfulLogin) {
-          Navigator.pushReplacementNamed(context, homeRoute);
+          final userResult = await AppUserSingleton().setUser();
```
Comment (lecicdjuro) [inline lib/presentation/login/view/login_page_view.dart:42]: not right place to fetch user. That should be part of bloc

**Review by lecicdjuro — CHANGES_REQUESTED** [review]

File: lib/data/firebase_firestore/firebase_db_manager.dart [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+  factory FirebaseDbManager() {
+    if (_instance == null) {
+      _instance = FirebaseDbManager.internal();
+    }
+    return _instance!;
+  }
+
+  static FirebaseDbManager get instance => FirebaseDbManager();
+
+  Future<String> AddBookImage(File? imageFile) async {
```
Comment (lecicdjuro) [inline lib/data/firebase_firestore/firebase_db_manager.dart:28]: method name should not start with capital letter

File: lib/presentation/book/bloc/home_page_bloc.dart [hunk 21]
```diff
@@ -1,20 +1,24 @@
...
+    on<AddBookImageEvent>(_onAddBookImage);
+    on<DeleteBookImage>(_onDeleteBookImage);
   }
 
   Future<void> _onBookAdded(
       AddNewBook event, Emitter<HomePageState> emit) async {
     emit(LoadingState());
 
-    final urlResult = await FirebaseDbManager.AddBookImage(event.imageFile);
+    final urlResult =
```
Comment (lecicdjuro) [inline lib/presentation/book/bloc/home_page_bloc.dart:20]: what if this call fails?

**Review by lecicdjuro — CHANGES_REQUESTED** [review]

**Review by milicaLits — COMMENTED** [review]
Check formatting.

File: lib/presentation/book/bloc/home_page_bloc.dart [hunk 22]
```diff
@@ -8,23 +8,33 @@ import 'package:flutter_bloc/flutter_bloc.dart';
...
 
   Future<void> _onBookAdded(
       AddNewBook event, Emitter<HomePageState> emit) async {
     emit(LoadingState());
 
-    final urlResult =
-        await FirebaseDbManager.internal().AddBookImage(event.imageFile);
+    try {
+      final urlResult =
+          await FirebaseDbManager.internal().uploadBookImage(event.imageFile);
```
Comment (petarLits) [inline lib/presentation/book/bloc/home_page_bloc.dart:23]: why internal()? we should call instance...

File: lib/presentation/book/bloc/home_page_bloc.dart [hunk 23]
```diff
@@ -8,23 +8,33 @@ import 'package:flutter_bloc/flutter_bloc.dart';
...
+      final Book? result = new Book(
+          author: event.author,
+          title: event.title,
+          imageUrl: urlResult,
+          id: '');
 
-    final Book? result = new Book(
-        author: event.author, title: event.title, imageUrl: urlResult, id: '');
+      await FirebaseDbManager().addBookToServer(result!);
```
Comment (petarLits) [inline lib/presentation/book/bloc/home_page_bloc.dart:30]: use instance

File: lib/models/book/book.dart [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+        author: '',
+        title: '',
+        imageUrl: '',
+        id: '',
+      );
+
+  Map<String, dynamic> toJson() => _$BookToJson(this);
+
+  @override
+  List<Object?> get props => [title, author, id, imageUrl, bookData];
```
Comment (petarLits) [inline lib/models/book/book.dart:36]: comma after last parameter, format...

**Review by petarLits — CHANGES_REQUESTED** [review]

File: lib/data/firebase_firestore/firebase_db_manager.dart [hunk 25]
```diff
@@ -25,7 +25,7 @@ class FirebaseDbManager {
 
   static FirebaseDbManager get instance => FirebaseDbManager();
 
-  Future<String> AddBookImage(File? imageFile) async {
+  Future<String> uploadBookImage(File? imageFile) async {
     final timeStamp = DateTime.now();
 
     final fileRef = storageRef.child('${timeStamp.toString()}');
```
Comment (milicaLits) [inline lib/data/firebase_firestore/firebase_db_manager.dart:31]: Use book title as title for directory where you will place image.  The same directory that will be used for placing book page images.

File: lib/data/firebase_firestore/firebase_db_manager.dart [hunk 26]
```diff
@@ -41,4 +41,25 @@ class FirebaseDbManager {
     });
     return Future.value(url);
   }
+
+  Future<void> addBookToServer(Book book) async {
+    final bookRef = db.collection("books").doc();
```
Comment (milicaLits) [inline lib/data/firebase_firestore/firebase_db_manager.dart:46]: you can add constants for this table names

File: lib/models/book/book_chapter.dart [hunk 27]
```diff
@@ -0,0 +1,21 @@
+import 'package:equatable/equatable.dart';
+import 'package:json_annotation/json_annotation.dart';
+
+part 'book_chapter.g.dart';
+
+class BookChapters extends Equatable {
+  BookChapters({this.chTitle = '', required this.chNumber});
```
Comment (milicaLits) [inline lib/models/book/book_chapter.dart:7]: Add comma, format

File: lib/models/book/book_data.dart [hunk 28]
```diff
@@ -0,0 +1,22 @@
...
+import 'package:json_annotation/json_annotation.dart';
+
+part 'book_data.g.dart';
+
+@JsonSerializable()
+class BookData extends Equatable {
+  List<BookChapters> chapters;
+  List<BookPages> pages;
+
+  BookData({required this.chapters, required this.pages});
```
Comment (milicaLits) [inline lib/models/book/book_data.dart:13]: Comma, format.
Constructor should be at top.

File: lib/presentation/book/bloc/home_page_bloc.dart [hunk 29]
```diff
@@ -8,23 +8,33 @@ import 'package:flutter_bloc/flutter_bloc.dart';
 class HomePageBloc extends Bloc<HomePageEvent, HomePageState> {
   HomePageBloc() : super((InitialState())) {
     on<AddNewBook>(_onBookAdded);
-    on<RefreshBooks>(_onRefreshBooks);
+    on<onNewBookAdded>(_onNewBookAdded);
```
Comment (milicaLits) [inline lib/presentation/book/bloc/home_page_bloc.dart:11]: For event name leave out "on". onNewBookAdded => NewBookAdded.

File: lib/presentation/book/bloc/home_page_bloc.dart [hunk 30]
```diff
@@ -1,5 +1,5 @@
 import 'package:book/data/firebase_firestore/firebase_db_manager.dart';
-import 'package:book/models/book.dart';
+import 'package:book/models/book/book.dart';
 import 'package:book/presentation/book/bloc/home_page_event.dart';
```
Comment (milicaLits) [inline lib/presentation/book/bloc/home_page_bloc.dart:3]: instead of importing files for events and state, you could use "part" and "part of".

**Review by milicaLits — CHANGES_REQUESTED** [review]

**Review by milicaLits — APPROVED** [review]

File: lib/presentation/book/bloc/home_page_event.dart [hunk 31]
```diff
@@ -24,13 +24,13 @@ class AddNewBook extends HomePageEvent {
       ];
 }
 
-class RefreshBooks extends HomePageEvent {
-  final List<Book>? bookList;
+class onNewBookAdded extends HomePageEvent {
```
Comment (lecicdjuro) [inline lib/presentation/book/bloc/home_page_event.dart]: Class name must start with capital letter

File: lib/presentation/book/bloc/home_page_bloc.dart (hunk 22 above)
Comment (lecicdjuro) [reply lib/presentation/book/bloc/home_page_bloc.dart]: internal method should be private method, not public

File: lib/presentation/book/bloc/home_page_bloc.dart [hunk 32]
```diff
@@ -1,15 +1,21 @@
...
+part 'home_page_event.dart';
+
+part 'home_page_state.dart';
+
 class HomePageBloc extends Bloc<HomePageEvent, HomePageState> {
   HomePageBloc() : super((InitialState())) {
     on<AddNewBook>(_onBookAdded);
-    on<onNewBookAdded>(_onNewBookAdded);
-    on<AddBookImageEvent>(_onAddBookImage);
+    on<NewBookAdded>(_NewBookAdded);
```
Comment (petarLits) [inline lib/presentation/book/bloc/home_page_bloc.dart:17]: methods should not start with capital letter

**Review by lecicdjuro — CHANGES_REQUESTED** [review]

**Review by petarLits — CHANGES_REQUESTED** [review]

File: lib/presentation/book/view/home_page_view.dart [hunk 33]
```diff
@@ -9,8 +16,149 @@ class HomePageView extends StatefulWidget {
...
+    return BlocConsumer<HomePageBloc, HomePageState>(
+      listener: (context, state) {
+        if (state is LoadingState) {
+          DialogUtils.showLoadingScreen(context);
+        } else if (state is LoadedState) {
+          Navigator.pop(context);
+        } else if (state is SuccessfulBookAdded) {
+          context.read<HomePageBloc>().add(DownloadBooks());
+        } else if (state is BooksDownloadedState) {
+          bookList.addAll(state.bookList);
```
Comment (lecicdjuro) [inline lib/presentation/book/view/home_page_view.dart:48]: this is a bug. You will always add books when this state happens. You need to clear bookList before you add. Or just bookList = List.from(state.bookList)

**Review by lecicdjuro — CHANGES_REQUESTED** [review]

**Review by petarLits — APPROVED** [review]

**Review by lecicdjuro — APPROVED** [review]

**Pull request merged**